### ISSUE TITLE
Arbitration locking issue

### ISSUE THREAD

**Issue by craigmiskell-gitlab** [tpitale/mail_room#84]
I think there's a problem with the redis arbitration locking, and we're seeing unread messages get stuck in our inbox as a result.  

The 'lock' is taken in deliver? which is called during the search for UNSEEN messages (new_message_ids).  If something goes wrong during the actual message fetch, then the message will not be marked SEEN.  When a mail_room calls new_message_ids (the restarted one, or another instance) before the key expires (10min by default), we'll call deliver? which will increment the key and reset the TTL, and return false (the value will != 1).  If we never go 10 minutes without receiving mail (or with no mail_room processes running), the key will keep on being incremented and the TTL keep on being reset, and the message will never be picked up again.  

I don't know what the solution is I'm afraid, but I'm happy to try and help figure it out.

**Comment by DouweM** [comment]
For reference, this is related to https://github.com/tpitale/mail_room/issues/46 and https://github.com/tpitale/mail_room/pull/49. Some related conversation (including a potential way forward) can also be found in https://gitlab.com/gitlab-org/gitlab-ce/issues/2870#note_4184125, where we discussed this from a GitLab perspective.

**Comment by tpitale** [comment]
Open to PRs! I believe gitlab is the primary user of arbitration in Redis.

**Comment by craigmiskell-gitlab** [comment]
Will work on a PR; pretty sure 

> SET key value NX EX timeout

from the end gitlab issue is the right way forward, but I'll run some tests to confirm.

**Comment by craigmiskell-gitlab** [comment]
TL;DR: it was the updating of the TTL at every INCR that was the problem, and SET with NX and an expiry solves that cleanly.  It's backwards compatible too (will stop updating the TTL, value will expire, and UNSEEN messages will then be processed).
